### PR TITLE
[AMD] [GLM5] Skip DSA decode indexer when kv_len <= index_topk (dense k-only fast path)

### PR DESCRIPTION
**Pull request by Jacob0226**
## Summary

- On GLM-5.2 DSA decode, when a request's `kv_len <= index_topk` the top-k selects **all** valid positions, so the indexer's logits GEMM + `paged_mqa_logits` + top-k selection is wasted work. Add a **k-only** fast path that skips the indexer, stores the K cache, and generates the identity index directly (`[0, 1, ..., kv_len-1, -1, ...]`), feeding the same sparse-MLA decode attention kernel.
- **CUDA-graph "Design A" dual-graph:** capture a dense (k-only) and a sparse (full indexer) decode graph per batch-size bucket and dispatch on `max_kv_len` vs `index_topk` at replay. This is correct for mixed lengths — any request with `kv_len > index_topk` in the batch falls back to the sparse graph.
- **Auto-enabled for DSA models, no env toggle.** Since DSA models expose `index_topk` in their HF config, the dual-graph is turned on automatically whenever `index_topk` is present (`dsa_dual_graph = dsa_index_topk is not None`); non-DSA models and other archs are unaffected. Eager decode takes the same fast path per-step when safe. (No `SGLANG_DSA_DECODE_*` / `SGLANG_KONLY_DEBUG` environment variables — an earlier revision gated this behind opt-in/debug flags; those were removed per review.)

## Changes

| File | Change |
| --- | --- |
| layers/attention/dsa/dsa_indexer.py | k-only fast path (`_forward_cuda_k_only`): skip logits/top-k, store K cache; for MLA generate the identity index; extended to DECODE (eager + graph) |
| model_executor/runner/decode_cuda_graph_runner.py | Design A dual-graph: capture dense (k-only) + sparse decode graphs per bs bucket; host-dispatch on `max_kv_len` vs `index_topk`; auto-enabled for DSA models via `dsa_index_topk` |
| model_executor/runner_utils/capture_mode.py | capture-time DSA variant flag (dense/sparse) read by the indexer skip branch |
| model_executor/runner/shape_key.py | include the DSA decode variant in the cuda-graph shape key |

## Scope

GLM-5.2 DSA decode on gfx950 (MI355X). Auto-enabled for DSA models; correct for mixed lengths (long context stays on the sparse indexer path). Other archs / hardware unaffected.

## Test plan

**Accuracy (GSM8K, MI355X TP4, GLM-5.2-MXFP4):**

| | Baseline | This PR |
| --- | --- | --- |
| GSM8K | 0.922 | 0.941 |

Within margin of error.

**Performance (MI355X TP4, GLM-5.2-MXFP4, docker `rocm/sgl-dev:v0.5.15.post1-rocm720-mi35x-20260714`, tilelang DSA backend):** token throughput per GPU (tok/s/gpu, higher better) and median TPOT (ms, lower better), dense-decode OFF vs ON (this PR). Isolated effect — same build/backend, only this PR's commits added, everything else identical.

Only **i1024 / o1024** is shown: this optimization applies when `kv_len <= index_topk` (2048). At i8192 every decode step has `kv_len > 2048`, so it stays on the sparse path and is unchanged.

| Concurrency | TPUT off | TPUT PR | Δ TPUT | TPOT off | TPOT PR | Δ TPOT |
| --- | --- | --- | --- | --- | --- | --- |
| 4  | 71.8  | 75.5  | +5.2% | 13.28 | 12.62 | −5.0% |
| 8  | 126.8 | 133.2 | +5.1% | 15.18 | 14.41 | −5.1% |
| 16 | 196.9 | 205.1 | +4.2% | 19.46 | 18.67 | −4.1% |
| 32 | 312.4 | 322.8 | +3.3% | 24.66 | 23.90 | −3.1% |
| 64 | 468.9 | 480.3 | +2.4% | 32.88 | 32.14 | −2.3% |

```
python3 -m sglang.launch_server \
  --model amd/GLM-5.2-MXFP4 \
  --tp 4 \
  --trust-remote-code \
  --tool-call-parser glm47 \
  --reasoning-parser glm45 \
  --watchdog-timeout 1200 \
  --mem-fraction-static 0.85 \
  --kv-cache-dtype fp8_e4m3 \
  --disable-radix-cache \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 32}' \
  --dsa-prefill-backend tilelang \
  --dsa-decode-backend tilelang \
  --tokenizer-worker-num 8 \
  --enable-aiter-allreduce-fusion
```

Token throughput per GPU improves ~2.4–5.2% and TPOT drops ~2.3–5.1% at i1k, from skipping the wasted indexer when the sparse top-k would select all positions anyway. The gain is largest at low concurrency and converges to ~2.4% at conc 64. (Δ is the isolated dense-decode effect — same build/backend, feature off vs on — not a full-stack-vs-stock comparison.)

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #31978978932](https://github.com/sgl-project/sglang/actions/runs/31978978932)
Latest PR Test (Extra): :x: [Run #31978978765](https://github.com/sgl-project/sglang/actions/runs/31978978765)

### CODE REVIEW

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 1]
```diff
@@ -534,12 +534,62 @@ def _maybe_rotate(self, x: torch.Tensor) -> torch.Tensor:
...
+                    return True
+                if variant == "sparse":
+                    return False
+
+                # M2a "static" fallback (no dual-variant capture active): enable
+                # k-only only when the deployment guarantees every request's
+                # kv_len<=index_topk (e.g. i1k/o1k). WRONG for mixed >2K traffic.
+                import os
+
+                return os.environ.get("SGLANG_DSA_DECODE_DENSE_GRAPH", "0") == "1"
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Done — switched to `envs.SGLANG_DSA_DECODE_DENSE_GRAPH.get()` in commit 474270e so it goes through the EnvBool parser (accepts `true`/`1`/`yes`).

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 2]
```diff
@@ -1302,8 +1352,19 @@ def _forward_cuda_k_only(
...
+        )
         x_meta = x[0] if isinstance(x, tuple) else x
+        if forward_batch.forward_mode.is_decode_or_idle() and layer_id == 0:
+            import os
+
+            if os.environ.get("SGLANG_KONLY_DEBUG", "0") == "1":
+                logger.info(
+                    "[KONLY] decode k-only fired (skip-indexer): num_tok=%d",
+                    x_meta.shape[0],
+                )
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Done — switched to `envs.SGLANG_KONLY_DEBUG.get()` in commit 474270e, same EnvBool parsing.

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 3]
```diff
@@ -534,12 +534,62 @@ def _maybe_rotate(self, x: torch.Tensor) -> torch.Tensor:
...
+        # cuda graph the chosen branch is frozen at capture time and would replay
+        # incorrectly for kv_len > index_topk, so decode skip is gated off during
+        # capture and handled separately by graph-variant / eager-fallback (M2).
+        fb = forward_batch
+
+        # Prefill/extend: original per-step gate (host sync on seq_lens_cpu is fine).
+        if fb.forward_mode.is_extend_without_speculative():
+            if fb.seq_lens_cpu is None:
+                return False
+            return int(fb.seq_lens_cpu.max().item()) <= self.index_topk
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py:409]: Done in 27bd826 — added the `fb.seq_lens_cpu is None or fb.seq_lens_cpu.numel() == 0` guard before `.max()`.

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 4]
```diff
@@ -534,12 +534,62 @@ def _maybe_rotate(self, x: torch.Tensor) -> torch.Tensor:
...
+                return os.environ.get("SGLANG_DSA_DECODE_DENSE_GRAPH", "0") == "1"
+            # Eager decode: safe to check per-step (host sync OK); correct for both
+            # kv_len<=index_topk (k-only) and kv_len>index_topk (falls through).
+            if fb.seq_lens_cpu is not None:
+                max_kv_len = int(fb.seq_lens_cpu.max().item())
+            elif fb.seq_lens is not None:
+                max_kv_len = int(fb.seq_lens.max().item())
+            else:
+                return False
             return max_kv_len <= self.index_topk
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py]: Done in 27bd826 — added `.numel() > 0` checks to both the `seq_lens_cpu` and `seq_lens` branches before `.max()`, consistent with the existing empty-batch guard elsewhere in this file.

**Review by clintg6 — CHANGES_REQUESTED** [review]
@Jacob0226 Tested on MI355X with GLM-5.2-MXFP4 (TP8, 1024 ISL / 1024 OSL). The dual-graph dispatch works correctly. Benchmark results showed around 3% uplift in Total token throughput. With that said, I have a number of recommendations to improve the PR.

## Remove SGLANG_KONLY_DEBUG

This appears to be a leftover debug flag. It's only used to gate a single logger.info call in _forward_cuda_k_only, so it shouldn't ship in production. Remove it from environ.py and the associated logging block in dsa_indexer.py.

## Remove SGLANG_DSA_DECODE_DUAL_GRAPH and auto-enable for DSA models
Since DSA models already expose index_topk in their HF config, users shouldn't need to enable this manually. Auto-detect DSA models instead:
```python
# Instead of:
self.dsa_dual_graph = envs.SGLANG_DSA_DECODE_DUAL_GRAPH.get()

# Use:
self.dsa_dual_graph = self.dsa_index_topk is not None
```

## Remove SGLANG_DSA_DECODE_DENSE_GRAPH

This forces dense decode by unconditionally bypassing the indexer, making it another debug/testing override and is unnecessary.

**Comment by Jacob0226** [comment]
Thanks @clintg6! Addressed all three in `442585399a`:
1. Removed `SGLANG_KONLY_DEBUG` (flag + debug logging/counters).
2. Removed `SGLANG_DSA_DECODE_DUAL_GRAPH` — now auto-enabled for DSA models via `dsa_dual_graph = dsa_index_topk is not None`.
3. Removed `SGLANG_DSA_DECODE_DENSE_GRAPH` — capture fallback defaults to the sparse path.
PR description updated too.

**Review by clintg6 — APPROVED** [review]
Tested latest changes and passed all performance and accuracy tests. Approving.

**Review by sogalin — COMMENTED** [review]
Have you tried the change on CUDA? It is common code change in SGL.

**Comment by Jacob0226** [comment]
> Have you tried the change on CUDA? It is common code change in SGL.

No. I add a commit to guard all codes under ROCm to prevent unexpected errors in upstream.

**Comment by Jacob0226** [comment]
@amd-bot ci-status

**Comment by Jacob0226** [comment]
image
CUDA CI (PR Test Base) ALL green

**Comment by Jacob0226** [comment]
image
AMD CI all green

**Review by sogalin — APPROVED** [review]
LGTM, having is_hip to protect now.

**Comment by Jacob0226** [comment]
@amd-bot ci-status

**Comment by Jacob0226** [comment]
@amd-bot ci-status

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
CI green, final update to comment https://github.com/sgl-project/sglang/pull/31324/commits/903fe7d8ea1fe6d934e0e88354678eb523849ae4
